hw/mcu/sifive/fe310/src/hal_spi.c

```diff
@@ -0,0 +1,384 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+#include <os/mynewt.h>
+#include <assert.h>
+
+#include <hal/hal_spi.h>
+#include <hal/hal_gpio.h>
+
+#include <defs/error.h>
+#include <syscfg/syscfg.h>
+
+#include <mcu/fe310_hal.h>
+#include <mcu/sys_clock.h>
+#include <mcu/plic.h>
+#include <sifive/devices/spi.h>
+#include <env/freedom-e300-hifive1/platform.h>
+
+struct fe310_hal_spi
+{
+    struct hal_spi_settings spi_cfg;
+
+    uint32_t spi_base;
+
+    int len;
+    int txleft;
+    int rxleft;
+
+    /* Pointers to tx/rx buffers */
+    uint8_t *txbuf;
+    uint8_t *rxbuf;
+
+    /* Callback and arguments */
+    hal_spi_txrx_cb txrx_cb_func;
+    void            *txrx_cb_arg;
+};
+
+#if MYNEWT_VAL(SPI_1)
+struct fe310_hal_spi fe310_hal_spi1 = {
+    .spi_base = SPI1_CTRL_ADDR
+};
+#endif
+
+#if MYNEWT_VAL(SPI_2)
+struct fe310_hal_spi fe310_hal_spi2 = {
+    .spi_base = SPI2_CTRL_ADDR
+};
+#endif
+
+static struct fe310_hal_spi *fe310_hal_spis[3] = {
+    NULL, /* SPI 0 used to access flash */
+#if MYNEWT_VAL(SPI_1)
+    &fe310_hal_spi1,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(SPI_2)
+    &fe310_hal_spi2,
+#else
+    NULL,
+#endif
+};
+
+#define FE310_HAL_SPI_RESOLVE(__n, __v)                     \
+    if ((__n) > 2) {                                        \
+        rc = SYS_EINVAL;                                    \
+        goto err;                                           \
+    }                                                       \
+    (__v) = (struct fe310_hal_spi *) fe310_hal_spis[(__n)]; \
+    if ((__v) == NULL) {                                    \
+        rc = SYS_EINVAL;                                    \
+        goto err;                                           \
+    }
+
+static void
+spi_interrupt_handler(int int_num)
+{
+    struct fe310_hal_spi *spi = fe310_hal_spis[int_num - INT_SPI0_BASE];
+    uint32_t val;
+    for (;;) {
```

Comment:
Is it possible to cap this to something? It would be best if we would have some limit on time spent within interrupt handler. I'm not familiar with this MCUs interrupt delivery mechanism, nor priorities. So I'm not sure if earlier return would gain anything, or if the system would get stuck without being able to any other work anyway.

Reply:
Added sanity cap